### ISSUE TITLE
Implement a periodic "pulse check" in order to maintain Verified status

### ISSUE THREAD

**Issue by CurlyBracketEffect** [BrightID/BrightID#824]
Right now once an account is created it will remain active indefinitely. So anyone wanting to create Sybils needs only to create an account, attend a party to get verified and then they can connect and ETH address and the apps they want to use (likely the ones that they can make money on), and then they can delete BrightID and use that ETH address with those apps indefinitely. They are then free to use the same device to repeat the process. 

If we could make it so a user needs to perform some basic task in the app within a set amount of time, and if they don't their BrightID goes into a dormant state, it would make it increasingly difficult to manage multiple bot/Sybil accounts. For example: Randomly once a week the app sends a push notification on the device, and the user has 24 hours in order to open the app and press the "I'm still human" button in order to keep their verification status. If they do not perform the task the BritghtID account will become "dormant", temporarily removing their Verified status so they cannot use the apps that they have linked. In order to gain the Verified status back, they will need to either:

1. Perform Social Recovery. 
2. Wait for another random push notification within 1 week and respond within 24 hours. 

If they miss the second notification or they do not do Social Recovery within 1 week, then they will have their Verified status permanently removed, and the only way to get it back would be to connect with a Seed Group once again. 

I feel like the weekly or bi-weekly action of a user would be easy and low barrier enough for an honest user, but becomes increasingly difficult for a malicious user. This would also start to give us some more metrics to use for Trust algorithms; If we see that a user is constantly responding to the notification within 24 hours this is a stronger sign of human uniqueness than if a user is constantly going into dormancy and having to recover their Verified status. It also adds a bit of gamefication where we can have a "Streak counter" and apps can use that to say an account needs to have been "active" (not become dormant) for X # of weeks in order to qualify for one of their apps offerings. 

One issue I could foresee with this is in an instance where a family or community only has one or a few phones to share amongst many people. I don't have a perfect solution for this, but I know we are working on a new connection classification like "family" (or something similar), that perhaps we can use to aid in these situations. However, I don't know enough about how that is proposed to work to suggest something right now. With the system designed as it is right now though it would already be difficult for a family sharing a phone to even do Social Recovery, which as I understand is something that we highly want to encourage as it will give better data for the Trust algorithms to use. So I think that is an issue we need to solve outside of this potential layer of Sybil defense.

**Comment by adamstallard** [comment]
I remember you bringing this up in the AMA.

I think one of the features I liked that you brought up then was that this would let users periodically re-link to the apps they care about by sending a "heartbeat" to those apps.

If you never open your BrightID app, it won't do it.

I think it could actually be a manual "heartbeat."  Like a button on the apps tab that says "Send heartbeat to 9 apps" and is enabled every 3 days or something.

If you don't do that, the apps can (and probably should) let your verification expire.

**Comment by bitsikka** [comment]
I like the idea of sourcing activity to filter out sybils. I'm not sure about the mechanisms though.

Ideally such an activity would be something natural that an individual does on their own accord rather than something that is subjected.

At the current state of the reference app, the activity mechanisms proposed, it seems, are the only kinds imaginable. However, perhaps in the future, if the app were to be enhanced to have more social features, then it is imaginable there could be more natural source of activities that could be used.

This all follows from the fact that BrightID is **socially derived identity**. It makes sense to me(if not now then in future) that we need to think about *social activities* going beyond "making connections or adding to contact list"(natural), "manually sending heartbeat"(less natural - enforced and not palatable).

There are activities (voting, transacting...etc) that happens in Apps that use BrightID but they are less frequent and detached from BrightId itself. Even those activities if they could be sourced back into BirghtID, it would be more natural and no-brainer(in the sense of UX).

My hypothesis is that, after making connections, *messaging* is perhaps the most natural and frequent social activity. Messaging is second to perhaps even more natural and frequent activity of "liking", "sharing/retweeting".

This is leading me to think about
- some kind of **proof-of-activity**(privacy preserving) protocol/standard feeding into a node's sybil rank.
- along with crypto native social platforms amenable to become the source of these natural heartbeat activities.

Problem is there isn't anything like that today - so unimaginable yet.

There are crypto native socials being built though that we should perhaps coordinate with.

**Comment by bitsikka** [comment]
@andrewhong5297 might have insights to share. Social [trans]actions were categorized in his mirror post on [The Composability of Identity across Web2 and Web3](https://ath.mirror.xyz/mDu_nGsWpjvF-KaPLADwxeAMfze3UJOtTAQi2CwPwWc). He is working on related standards.

**Comment by andrewhong5297** [comment]
Thanks @bitsikka , yeah I do think there is a way of managing this based on transactions. Simon Taylor brought this up in a [response to my article](https://twitter.com/sytaylor/status/1404532773424148482), mentioning some DARPA work that allows you to detect a unique actor with only 3-4 transactions. I'm starting some work on exporting/webscraping social graphs in web3, starting with mirror.xyz/race participants since I can see how the "nodes" were voted in as well as their ethereum and twitter address. I do think web3 actions + web2 actions + clustering of social graph nodes into a model should give strong predictive power to uniqueness. My project is working on [collecting these web3 + web2 identifiers into the wallet ](https://github.com/andrewhong5297/identity-hack) as a first phase, then outwards into enabling different use cases like this one. 

tldr; layering platforms on top of social graphs could be the next step, and some classification or anomaly detection model could work well.

**Comment by bitsikka** [comment]
Thanks @andrewhong5297 :) fascinating stuff. That DARPA work seems like a really awesome thing to look into for Sybil resistant id like BrightiD + your project work. Please point to the resource when you find it. I am definitely curious about how it can be leveraged in a privacy-preserving/unintrusive/opt-in/self-sovereign manner.

> layering platforms on top of social graphs could be the next step, and some classification or anomaly detection model could work well.

most definitely 👍🏽

**Comment by CurlyBracketEffect** [comment]
> I think it could actually be a manual "heartbeat." Like a button on the apps tab that says "Send heartbeat to 9 apps" and is enabled every 3 days or something.
> 
> If you don't do that, the apps can (and probably should) let your verification expire.

Yes, my proposal is more along the lines of what you said here @adamstallard, but I think every 3 days might be a little too frequent. The reason why I suggested that it occur randomly once a week is that I feel this should be as infrequent as possible so as to not be completely annoying to non-malicious users, but frequent enough, and not predictive so that it is really annoying for malicious users. The more Sybil accounts you are trying to manage the harder it will be to maintain all of their "heartbeats". By making it non-predictive, one would have to log into a BrightID account more frequently to check if they need to perform a heartbeat check, but if you are one user with one phone, then the push notification will always come through and you just need to log in that day to complete the heartbeat check. 

> Ideally such an activity would be something natural that an individual does on their own accord rather than something that is subjected.

To @bitsikka's point here, I would like to see this be tied to the "natural" behaviour of a user, but think that it is crucial that this require performing some task within the mobile app itself though. I don't like the idea of it being based on transaction history because: 

1. We start getting into privacy issues and people will be more resistant to that. Without looking at the technical implementation too closely, it sounds like DARPA addresses the privacy issue, but read on....
2. Being open-source it would be easy enough for a Sybil puppet master to just create another bot that performs all the transactions and Twitter actions necessary in order to fulfil the "heartbeat" requirements. If it wasn't open-source it would still just be a matter of time before someone figured it out.
3. Who would decide what transactions are worthy? If the expense of performing these transactions are too high then it isn't profitable for a user to use BrightID, and if they are too low then again it's still viable for a Sybil to game. 
4. Suppose a person wants to use a single (or multiple) eth address(es) for all their BrightID related tasks, and they do all of their other actions which would qualify as "heartbeat actions", with another address. We would need a way to consolidate all of this into one useable "heartbeat" verification. DARPA might be able to satisfy this as well, but it also sounds like the project is still being worked on and is not ready for the use case I am proposing right now. 

> My hypothesis is that, after making connections, _messaging_ is perhaps the most natural and frequent social activity. Messaging is second to perhaps even more natural and frequent activity of "liking", "sharing/retweeting".

I am not a fan of adding messaging to the app because:
1. We already have far too many messaging apps out there
2. We would get into the same issue as using transaction history data where we need to ensure people's messaging data is secure.
3. Again, if we are using a third-party application outside the app, it would be easier to write bots that just chat with each other. 

I also think it's worth mentioning that none of the actions that people currently use BrightID for right now is "natural behaviour" unless you count performing an action to receive a reward as natural. The most used apps are giving out rewards for people to perform an action and were it not for those rewards nobody would be performing the actions at all. 

The heartbeat action I am proposing benefits the BrightID network as a whole by increasing Sybil resistance and as a result, I think it would be worth providing some reward. Perhaps a gamefied tiered reward system where the user would receive a graduating amount of tokens in return for maintaining a longer and longer "heartbeat" streak. This would be in line with the current action=>reward model that most apps are using BrightID for right now. 

> @andrewhong5297 might have insights to share. Social [trans]actions were categorized in his mirror post on [The Composability of Identity across Web2 and Web3](https://ath.mirror.xyz/mDu_nGsWpjvF-KaPLADwxeAMfze3UJOtTAQi2CwPwWc). He is working on related standards.

Having read over this brief, I think DARPA sounds like a really cool project and look forward to seeing how it develops. It seems like it will act as the glue to bind many disparate parts that make up an "identity" into a useful primative. It seems like more work needs to be done though before we could use it in BrightID. As BrightID is listed as a part of the model I am interested to see how it will recursively feedback into the system were we to use it for the purposes I am describing. In the model, it is using BrightID as a piece of Sybil resistance, but I am suggesting something that would increase Sybil resistance, so what happens when we use a system that uses our system to also increase Sybil resistance? Sounds like Sybil resistance inception...which sounds so crazy it just might work, but is still unclear. 

All in all, DARPA sounds like a high;y technical potential solution to the ever-evolving problem of Sybil resistance. However, it's not yet ready to be tried out today. What I am proposing is far less technical and we have the tools to begin implementing it today. This would also make the Sybil resistance BrightID provides even strong to then be used in DARPA at any stage.

**Comment by andrewhong5297** [comment]
Just to clarify DARPA is separate from. My project, and also the branches is not a final method or organization yet, more to come on that soon! But yeah its no where near the point where brightID can use it yet.

**Comment by CurlyBracketEffect** [comment]
> Just to clarify DARPA is separate from my project, and also the branches is not a final method or organization yet, more to come on that soon! But yeah its no where near the point where brightID can use it yet.

Thanks for that clarification, sorry for the conflation.

**Comment by andrewhong5297** [comment]
no problem! 

> The heartbeat action I am proposing benefits the BrightID network as a whole by increasing Sybil resistance and as a result, I think it would be worth providing some reward.

I do like this idea
